tools/document-pii/src/pii_finder.rs

```diff
@@ -0,0 +1,266 @@
+        });
+
+        if has_pii_value(self.pii_values, node) {
+            self.pii_types.insert(self.current_path.clone());
+        }
+
+        self.visit_field_types(node);
+
+        self.current_path.pop();
+    }
+}
+
+/// if you have a field such as `foo: Foo<Bar<Baz>>` this function can take the type of the field
+/// and return a vector of the types like: ["Foo", "Bar", "Baz"].
+fn get_field_types(ty: &Type, segments: &mut Vec<String>) {
+    match ty {
+        Type::Path(type_path) => {
+            let mut path_iter = type_path.path.segments.iter();
+            let first_segment = path_iter.next();
+
+            if let Some(first_segment) = first_segment {
+                let mut ident = first_segment.ident.to_string();
+
+                let args = &first_segment.arguments;
+                if let syn::PathArguments::AngleBracketed(angle_bracketed) = args {
+                    for generic_arg in angle_bracketed.args.iter() {
+                        match generic_arg {
+                            syn::GenericArgument::Type(ty) => {
+                                get_field_types(ty, segments);
+                            }
+                            _ => continue,
+                        }
+                    }
+                }
+
+                if let Some(second_segment) = path_iter.next() {
+                    ident.push_str("::");
+                    ident.push_str(&second_segment.ident.to_string());
+                    segments.push(ident);
+                } else {
+                    segments.push(ident);
+                }
+            }
+        }
+        _ => {
+            use quote::ToTokens;
+            let mut tokens = proc_macro2::TokenStream::new();
+            ty.to_tokens(&mut tokens);
+            segments.push(tokens.to_string());
+        }
+    }
+}
+
+/// Checks if an attribute is equal to a specified name and value.
+fn has_attr_value(attr: &Attribute, name: &str, value: &str) -> bool {
+    if let Meta::List(meta_list) = &attr.meta {
+        if meta_list.path.is_ident("metastructure") {
+            let mut iter = meta_list.tokens.clone().into_iter();
+            while let Some(token) = iter.next() {
+                if let proc_macro2::TokenTree::Ident(ident) = &token {
+                    if ident == name {
+                        if let Some(proc_macro2::TokenTree::Punct(punct)) = iter.next() {
+                            if punct.as_char() == '=' {
+                                if let Some(proc_macro2::TokenTree::Literal(lit_val)) = iter.next()
+                                {
+                                    let mut lit_val = lit_val.to_string();
+                                    lit_val.pop(); // remove superfluous quotes. "\"true\"" -> "true"
+                                    lit_val.remove(0);
+                                    if lit_val == value {
+                                        return true;
```

Comment:
Here we have 9 indentation levels, could we split the function into multiple so that it's more readable?

Reply:
ok fixed, it's now flatter than denmark